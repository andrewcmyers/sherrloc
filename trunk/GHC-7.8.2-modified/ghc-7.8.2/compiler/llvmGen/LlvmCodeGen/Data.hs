-- ----------------------------------------------------------------------------
-- | Handle conversion of CmmData to LLVM code.
--

module LlvmCodeGen.Data (
        genLlvmData
    ) where

#include "HsVersions.h"

import Llvm
import LlvmCodeGen.Base

import BlockId
import CLabel
import Cmm

import FastString
import qualified Outputable

-- ----------------------------------------------------------------------------
-- * Constants
--

-- | The string appended to a variable name to create its structure type alias
structStr :: LMString
structStr = fsLit "_struct"

-- ----------------------------------------------------------------------------
-- * Top level
--

-- | Pass a CmmStatic section to an equivalent Llvm code.
genLlvmData :: (Section, CmmStatics) -> LlvmM LlvmData
genLlvmData (sec, Statics lbl xs) = do
    label <- strCLabel_llvm lbl
    static <- mapM genData xs
    let types   = map getStatType static

        strucTy = LMStruct types
        alias   = LMAlias ((label `appendFS` structStr), strucTy)

        struct         = Just $ LMStaticStruc static alias
        link           = if (externallyVisibleCLabel lbl)
                            then ExternallyVisible else Internal
        const          = if isSecConstant sec then Constant else Global
        glob           = LMGlobalVar label alias link Nothing Nothing const

    return ([LMGlobal glob struct], [alias])

-- | Should a data in this section be considered constant
isSecConstant :: Section -> Bool
isSecConstant Text                    = True
isSecConstant ReadOnlyData            = True
isSecConstant RelocatableReadOnlyData = True
isSecConstant ReadOnlyData16          = True
isSecConstant Data                    = False
isSecConstant UninitialisedData       = False
isSecConstant (OtherSection _)        = False


-- ----------------------------------------------------------------------------
-- * Generate static data
--

-- | Handle static data
genData :: CmmStatic -> LlvmM LlvmStatic

genData (CmmString str) = do
    let v  = map (\x -> LMStaticLit $ LMIntLit (fromIntegral x) i8) str
        ve = v ++ [LMStaticLit $ LMIntLit 0 i8]
    return $ LMStaticArray ve (LMArray (length ve) i8)

genData (CmmUninitialised bytes)
    = return $ LMUninitType (LMArray bytes i8)

genData (CmmStaticLit lit)
    = genStaticLit lit

-- | Generate Llvm code for a static literal.
--
-- Will either generate the code or leave it unresolved if it is a 'CLabel'
-- which isn't yet known.
genStaticLit :: CmmLit -> LlvmM LlvmStatic
genStaticLit (CmmInt i w)
    = return $ LMStaticLit (LMIntLit i (LMInt $ widthInBits w))

genStaticLit (CmmFloat r w)
    = return $ LMStaticLit (LMFloatLit (fromRational r) (widthToLlvmFloat w))

genStaticLit (CmmVec ls)
    = do sls <- mapM toLlvmLit ls
         return $ LMStaticLit (LMVectorLit sls)
  where
    toLlvmLit :: CmmLit -> LlvmM LlvmLit
    toLlvmLit lit = do
      slit <- genStaticLit lit
      case slit of
        LMStaticLit llvmLit -> return llvmLit
        _ -> panic "genStaticLit"

-- Leave unresolved, will fix later
genStaticLit cmm@(CmmLabel l) = do
    var <- getGlobalPtr =<< strCLabel_llvm l
    dflags <- getDynFlags
    let ptr = LMStaticPointer var
        lmty = cmmToLlvmType $ cmmLitType dflags cmm
    return $ LMPtoI ptr lmty

genStaticLit (CmmLabelOff label off) = do
    dflags <- getDynFlags
    var <- genStaticLit (CmmLabel label)
    let offset = LMStaticLit $ LMIntLit (toInteger off) (llvmWord dflags)
    return $ LMAdd var offset

genStaticLit (CmmLabelDiffOff l1 l2 off) = do
    dflags <- getDynFlags
    var1 <- genStaticLit (CmmLabel l1)
    var2 <- genStaticLit (CmmLabel l2)
    let var = LMSub var1 var2
        offset = LMStaticLit $ LMIntLit (toInteger off) (llvmWord dflags)
    return $ LMAdd var offset

genStaticLit (CmmBlock b) = genStaticLit $ CmmLabel $ infoTblLbl b

genStaticLit (CmmHighStackMark)
    = panic "genStaticLit: CmmHighStackMark unsupported!"

-- -----------------------------------------------------------------------------
-- * Misc
--

-- | Error Function
panic :: String -> a
panic s = Outputable.panic $ "LlvmCodeGen.Data." ++ s

