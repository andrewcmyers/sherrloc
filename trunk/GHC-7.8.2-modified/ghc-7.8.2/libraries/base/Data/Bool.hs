{-# LANGUAGE Trustworthy #-}
{-# LANGUAGE NoImplicitPrelude #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  Data.Bool
-- Copyright   :  (c) The University of Glasgow 2001
-- License     :  BSD-style (see the file libraries/base/LICENSE)
-- 
-- Maintainer  :  libraries@haskell.org
-- Stability   :  experimental
-- Portability :  portable
--
-- The 'Bool' type and related functions.
--
-----------------------------------------------------------------------------

module Data.Bool (
   -- * Booleans
   Bool(..),
   -- ** Operations 
   (&&),
   (||),
   not,
   otherwise,
   bool,
  ) where

import GHC.Base

-- | Case analysis for the 'Bool' type.
-- @bool a b p@ evaluates to @a@ when @p@ is @False@, and evaluates to @b@
-- when @p@ is @True@.
--
-- /Since: 4.7.0.0/
bool :: a -> a -> Bool -> a
bool f _ False = f
bool _ t True  = t
