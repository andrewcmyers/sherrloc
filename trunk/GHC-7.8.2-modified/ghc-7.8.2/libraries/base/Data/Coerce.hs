{-# LANGUAGE Unsafe #-}
{-# LANGUAGE DeriveDataTypeable #-}
{-# LANGUAGE StandaloneDeriving #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  Data.Coerce
-- Copyright   :  (c) The University of Glasgow 2001
-- License     :  BSD-style (see the file libraries/base/LICENSE)
--
-- Maintainer  :  libraries@haskell.org
-- Stability   :  provisional
-- Portability :  portable
--
-- Safe coercions between data types.
--
-- More in-depth information can be found on the
-- <https://ghc.haskell.org/trac/ghc/wiki/Roles Roles wiki page>
--
-- /Since: 4.7.0.0/
-----------------------------------------------------------------------------

module Data.Coerce
        ( -- * Safe coercions
          coerce, Coercible,
        ) where
import GHC.Prim (coerce, Coercible)
