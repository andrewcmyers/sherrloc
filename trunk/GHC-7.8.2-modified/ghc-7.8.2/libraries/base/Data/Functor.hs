{-# LANGUAGE Trustworthy #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  Data.Functor
-- Copyright   :  (c) The University of Glasgow 2001
-- License     :  BSD-style (see the file libraries/base/LICENSE)
-- 
-- Maintainer  :  libraries@haskell.org
-- Stability   :  provisional
-- Portability :  portable
--
-- Functors: uniform action over a parameterized type, generalizing the
-- 'map' function on lists.

module Data.Functor
    (
      Functor(fmap),
      (<$),
      ($>),
      (<$>),
      void,
    ) where

import Control.Monad
import GHC.Base (Functor(..))

infixl 4 <$>

-- | An infix synonym for 'fmap'.
(<$>) :: Functor f => (a -> b) -> f a -> f b
(<$>) = fmap

infixl 4 $>

-- | Flipped version of '<$'.
--
-- /Since: 4.7.0.0/
($>) :: Functor f => f a -> b -> f b
($>) = flip (<$)

