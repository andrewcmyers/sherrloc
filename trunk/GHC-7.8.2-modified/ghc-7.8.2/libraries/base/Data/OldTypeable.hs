{-# LANGUAGE Trustworthy #-}
{-# LANGUAGE CPP
           , NoImplicitPrelude
           , OverlappingInstances
           , ScopedTypeVariables
           , FlexibleInstances
  #-}
{-# OPTIONS_GHC -funbox-strict-fields -fno-warn-warnings-deprecations #-}

-- The -XOverlappingInstances flag allows the user to over-ride
-- the instances for Typeable given here.  In particular, we provide an instance
--      instance ... => Typeable (s a) 
-- But a user might want to say
--      instance ... => Typeable (MyType a b)

-----------------------------------------------------------------------------
-- |
-- Module      :  Data.Typeable
-- Copyright   :  (c) The University of Glasgow, CWI 2001--2004
-- License     :  BSD-style (see the file libraries/base/LICENSE)
-- 
-- Maintainer  :  libraries@haskell.org
-- Stability   :  experimental
-- Portability :  portable
--
-- This module defines the old, kind-monomorphic 'Typeable' class. It is now
-- deprecated; users are recommended to use the kind-polymorphic
-- "Data.Typeable" module instead.
--
-- /Since: 4.7.0.0/
-----------------------------------------------------------------------------

module Data.OldTypeable {-# DEPRECATED "Use Data.Typeable instead" #-} -- deprecated in 7.8
  (

        -- * The Typeable class
        Typeable( typeOf ),     -- :: a -> TypeRep

        -- * Type-safe cast
        cast,                   -- :: (Typeable a, Typeable b) => a -> Maybe b
        gcast,                  -- a generalisation of cast

        -- * Type representations
        TypeRep,        -- abstract, instance of: Eq, Show, Typeable
        showsTypeRep,

        TyCon,          -- abstract, instance of: Eq, Show, Typeable
        tyConString,    -- :: TyCon   -> String
        tyConPackage,   -- :: TyCon   -> String
        tyConModule,    -- :: TyCon   -> String
        tyConName,      -- :: TyCon   -> String

        -- * Construction of type representations
        mkTyCon,        -- :: String  -> TyCon
        mkTyCon3,       -- :: String  -> String -> String -> TyCon
        mkTyConApp,     -- :: TyCon   -> [TypeRep] -> TypeRep
        mkAppTy,        -- :: TypeRep -> TypeRep   -> TypeRep
        mkFunTy,        -- :: TypeRep -> TypeRep   -> TypeRep

        -- * Observation of type representations
        splitTyConApp,  -- :: TypeRep -> (TyCon, [TypeRep])
        funResultTy,    -- :: TypeRep -> TypeRep   -> Maybe TypeRep
        typeRepTyCon,   -- :: TypeRep -> TyCon
        typeRepArgs,    -- :: TypeRep -> [TypeRep]
        typeRepKey,     -- :: TypeRep -> IO TypeRepKey
        TypeRepKey,     -- abstract, instance of Eq, Ord

        -- * The other Typeable classes
        -- | /Note:/ The general instances are provided for GHC only.
        Typeable1( typeOf1 ),   -- :: t a -> TypeRep
        Typeable2( typeOf2 ),   -- :: t a b -> TypeRep
        Typeable3( typeOf3 ),   -- :: t a b c -> TypeRep
        Typeable4( typeOf4 ),   -- :: t a b c d -> TypeRep
        Typeable5( typeOf5 ),   -- :: t a b c d e -> TypeRep
        Typeable6( typeOf6 ),   -- :: t a b c d e f -> TypeRep
        Typeable7( typeOf7 ),   -- :: t a b c d e f g -> TypeRep
        gcast1,                 -- :: ... => c (t a) -> Maybe (c (t' a))
        gcast2,                 -- :: ... => c (t a b) -> Maybe (c (t' a b))

        -- * Default instances
        -- | /Note:/ These are not needed by GHC, for which these instances
        -- are generated by general instance declarations.
        typeOfDefault,  -- :: (Typeable1 t, Typeable a) => t a -> TypeRep
        typeOf1Default, -- :: (Typeable2 t, Typeable a) => t a b -> TypeRep
        typeOf2Default, -- :: (Typeable3 t, Typeable a) => t a b c -> TypeRep
        typeOf3Default, -- :: (Typeable4 t, Typeable a) => t a b c d -> TypeRep
        typeOf4Default, -- :: (Typeable5 t, Typeable a) => t a b c d e -> TypeRep
        typeOf5Default, -- :: (Typeable6 t, Typeable a) => t a b c d e f -> TypeRep
        typeOf6Default  -- :: (Typeable7 t, Typeable a) => t a b c d e f g -> TypeRep

  ) where

import Data.OldTypeable.Internal hiding (mkTyCon)

import Unsafe.Coerce
import Data.Maybe

import GHC.Base

import GHC.Fingerprint.Type
import GHC.Fingerprint

#include "OldTypeable.h"

{-# DEPRECATED typeRepKey "TypeRep itself is now an instance of Ord" #-} -- deprecated in 7.2
-- | (DEPRECATED) Returns a unique key associated with a 'TypeRep'.
-- This function is deprecated because 'TypeRep' itself is now an
-- instance of 'Ord', so mappings can be made directly with 'TypeRep'
-- as the key.
--
typeRepKey :: TypeRep -> IO TypeRepKey
typeRepKey (TypeRep f _ _) = return (TypeRepKey f)

        -- 
        -- let fTy = mkTyCon "Foo" in show (mkTyConApp (mkTyCon ",,")
        --                                 [fTy,fTy,fTy])
        -- 
        -- returns "(Foo,Foo,Foo)"
        --
        -- The TypeRep Show instance promises to print tuple types
        -- correctly. Tuple type constructors are specified by a 
        -- sequence of commas, e.g., (mkTyCon ",,,,") returns
        -- the 5-tuple tycon.

newtype TypeRepKey = TypeRepKey Fingerprint
  deriving (Eq,Ord)

----------------- Construction ---------------------

{-# DEPRECATED mkTyCon "either derive Typeable, or use mkTyCon3 instead" #-} -- deprecated in 7.2
-- | Backwards-compatible API
mkTyCon :: String       -- ^ unique string
        -> TyCon        -- ^ A unique 'TyCon' object
mkTyCon name = TyCon (fingerprintString name) "" "" name

-------------------------------------------------------------
--
--              Type-safe cast
--
-------------------------------------------------------------

-- | The type-safe cast operation
cast :: (Typeable a, Typeable b) => a -> Maybe b
cast x = r
       where
         r = if typeOf x == typeOf (fromJust r)
               then Just $ unsafeCoerce x
               else Nothing

-- | A flexible variation parameterised in a type constructor
gcast :: (Typeable a, Typeable b) => c a -> Maybe (c b)
gcast x = r
 where
  r = if typeOf (getArg x) == typeOf (getArg (fromJust r))
        then Just $ unsafeCoerce x
        else Nothing
  getArg :: c x -> x 
  getArg = undefined

-- | Cast for * -> *
gcast1 :: (Typeable1 t, Typeable1 t') => c (t a) -> Maybe (c (t' a)) 
gcast1 x = r
 where
  r = if typeOf1 (getArg x) == typeOf1 (getArg (fromJust r))
       then Just $ unsafeCoerce x
       else Nothing
  getArg :: c x -> x 
  getArg = undefined

-- | Cast for * -> * -> *
gcast2 :: (Typeable2 t, Typeable2 t') => c (t a b) -> Maybe (c (t' a b)) 
gcast2 x = r
 where
  r = if typeOf2 (getArg x) == typeOf2 (getArg (fromJust r))
       then Just $ unsafeCoerce x
       else Nothing
  getArg :: c x -> x 
  getArg = undefined

