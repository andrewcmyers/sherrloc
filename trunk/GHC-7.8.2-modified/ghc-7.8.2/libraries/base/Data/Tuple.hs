{-# LANGUAGE Trustworthy #-}
{-# LANGUAGE NoImplicitPrelude #-}
{-# OPTIONS_GHC -fno-warn-unused-imports #-}
-- XXX -fno-warn-unused-imports needed for the GHC.Tuple import below. Sigh.

-----------------------------------------------------------------------------
-- |
-- Module      :  Data.Tuple
-- Copyright   :  (c) The University of Glasgow 2001
-- License     :  BSD-style (see the file libraries/base/LICENSE)
-- 
-- Maintainer  :  libraries@haskell.org
-- Stability   :  experimental
-- Portability :  portable
--
-- The tuple data types, and associated functions.
--
-----------------------------------------------------------------------------

module Data.Tuple
  ( fst
  , snd
  , curry
  , uncurry
  , swap
  )
    where

import GHC.Base
-- We need to depend on GHC.Base so that
-- a) so that we get GHC.Classes, GHC.Types

-- b) so that GHC.Base.inline is available, which is used
--    when expanding instance declarations

import GHC.Tuple
-- We must import GHC.Tuple, to ensure sure that the
-- data constructors of `(,)' are in scope when we do
-- the standalone deriving instance for Eq (a,b) etc

default ()              -- Double isn't available yet

-- ---------------------------------------------------------------------------
-- Standard functions over tuples

-- | Extract the first component of a pair.
fst                     :: (a,b) -> a
fst (x,_)               =  x

-- | Extract the second component of a pair.
snd                     :: (a,b) -> b
snd (_,y)               =  y

-- | 'curry' converts an uncurried function to a curried function.
curry                   :: ((a, b) -> c) -> a -> b -> c
curry f x y             =  f (x, y)

-- | 'uncurry' converts a curried function to a function on pairs.
uncurry                 :: (a -> b -> c) -> ((a, b) -> c)
uncurry f p             =  f (fst p) (snd p)

-- | Swap the components of a pair.
swap                    :: (a,b) -> (b,a)
swap (a,b)              = (b,a)
