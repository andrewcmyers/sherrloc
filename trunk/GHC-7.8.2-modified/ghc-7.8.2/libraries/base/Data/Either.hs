{-# LANGUAGE Trustworthy #-}
{-# LANGUAGE NoImplicitPrelude #-}
{-# LANGUAGE DeriveDataTypeable, StandaloneDeriving #-}
{-# LANGUAGE PolyKinds, DataKinds, TypeFamilies, TypeOperators, UndecidableInstances #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  Data.Either
-- Copyright   :  (c) The University of Glasgow 2001
-- License     :  BSD-style (see the file libraries/base/LICENSE)
-- 
-- Maintainer  :  libraries@haskell.org
-- Stability   :  experimental
-- Portability :  portable
--
-- The Either type, and associated operations.
--
-----------------------------------------------------------------------------

module Data.Either (
   Either(..),
   either,
   lefts,
   rights,
   isLeft,
   isRight,
   partitionEithers,
 ) where

import GHC.Base
import GHC.Show
import GHC.Read

import Data.Typeable
import Data.Type.Equality

{-
-- just for testing
import Test.QuickCheck
-}

{-|

The 'Either' type represents values with two possibilities: a value of
type @'Either' a b@ is either @'Left' a@ or @'Right' b@.

The 'Either' type is sometimes used to represent a value which is
either correct or an error; by convention, the 'Left' constructor is
used to hold an error value and the 'Right' constructor is used to
hold a correct value (mnemonic: \"right\" also means \"correct\").
-}
data  Either a b  =  Left a | Right b
  deriving (Eq, Ord, Read, Show, Typeable)

instance Functor (Either a) where
    fmap _ (Left x) = Left x
    fmap f (Right y) = Right (f y)

instance Monad (Either e) where
    return = Right
    Left  l >>= _ = Left l
    Right r >>= k = k r

-- | Case analysis for the 'Either' type.
-- If the value is @'Left' a@, apply the first function to @a@;
-- if it is @'Right' b@, apply the second function to @b@.
either                  :: (a -> c) -> (b -> c) -> Either a b -> c
either f _ (Left x)     =  f x
either _ g (Right y)    =  g y

-- | Extracts from a list of 'Either' all the 'Left' elements
-- All the 'Left' elements are extracted in order.

lefts   :: [Either a b] -> [a]
lefts x = [a | Left a <- x]

-- | Extracts from a list of 'Either' all the 'Right' elements
-- All the 'Right' elements are extracted in order.

rights   :: [Either a b] -> [b]
rights x = [a | Right a <- x]

-- | Partitions a list of 'Either' into two lists
-- All the 'Left' elements are extracted, in order, to the first
-- component of the output.  Similarly the 'Right' elements are extracted
-- to the second component of the output.

partitionEithers :: [Either a b] -> ([a],[b])
partitionEithers = foldr (either left right) ([],[])
 where
  left  a ~(l, r) = (a:l, r)
  right a ~(l, r) = (l, a:r)

-- | Return `True` if the given value is a `Left`-value, `False` otherwise.
--
-- /Since: 4.7.0.0/
isLeft :: Either a b -> Bool
isLeft (Left  _) = True
isLeft (Right _) = False

-- | Return `True` if the given value is a `Right`-value, `False` otherwise.
--
-- /Since: 4.7.0.0/
isRight :: Either a b -> Bool
isRight (Left  _) = False
isRight (Right _) = True

-- instance for the == Boolean type-level equality operator
type family EqEither a b where
  EqEither (Left x)  (Left y)  = x == y
  EqEither (Right x) (Right y) = x == y
  EqEither a         b         = False
type instance a == b = EqEither a b

{-
{--------------------------------------------------------------------
  Testing
--------------------------------------------------------------------}
prop_partitionEithers :: [Either Int Int] -> Bool
prop_partitionEithers x =
  partitionEithers x == (lefts x, rights x)
-}

