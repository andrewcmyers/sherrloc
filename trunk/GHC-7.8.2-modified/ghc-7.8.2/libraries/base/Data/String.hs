{-# LANGUAGE Trustworthy #-}
{-# LANGUAGE NoImplicitPrelude, FlexibleInstances #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  Data.String
-- Copyright   :  (c) The University of Glasgow 2007
-- License     :  BSD-style (see the file libraries/base/LICENSE)
--
-- Maintainer  :  libraries@haskell.org
-- Stability   :  experimental
-- Portability :  portable
--
-- The @String@ type and associated operations.
--
-----------------------------------------------------------------------------

module Data.String (
   String
 , IsString(..)

 -- * Functions on strings
 , lines
 , words
 , unlines
 , unwords
 ) where

import GHC.Base
import Data.List (lines, words, unlines, unwords)

-- | Class for string-like datastructures; used by the overloaded string
--   extension (-XOverloadedStrings in GHC).
class IsString a where
    fromString :: String -> a

instance IsString [Char] where
    fromString xs = xs

