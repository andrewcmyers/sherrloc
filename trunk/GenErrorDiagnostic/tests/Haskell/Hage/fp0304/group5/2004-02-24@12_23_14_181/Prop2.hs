module Prop2 where

import Data.List

elemBy        :: (a -> a -> Bool) -> a -> [a] -> Bool
elemBy = undefined
eqString      :: String -> String -> Bool 
eqString = undefined
eqList        :: (a -> a -> Bool) -> [a] -> [a] -> Bool
eqList = undefined
eqChar      :: Char -> Char -> Bool
eqChar = undefined
eqBool      :: Bool -> Bool -> Bool 
eqBool = undefined

data Prop
      = En [Prop]
      | Of [Prop]
      | Niet Prop
      | Prop :-> Prop
      | Var String
      | Bool Bool

type Bedeling = [String]










evalueer :: Prop -> Bedeling -> Bool
evalueer (Bool b) _ = b
evalueer (Var _) [] = False
evalueer (Var s) (x:xs) | eqString s x = True
                        | otherwise = evalueer (Var s) xs
evalueer (En (p:ps)) xs = evalueer p xs && evalueer (En ps) xs
evalueer (En _ ) _ = True
evalueer (Of (p:ps)) xs = evalueer p xs || evalueer (Of ps) xs
evalueer (Of _ ) _ = False
evalueer (Niet ps) xs = not (evalueer ps xs)
evalueer (p :-> q) xs = not (evalueer p xs) || evalueer q xs










vervulbaar :: Prop -> [Bedeling]
vervulbaar prop = filter (evalueer prop) (geefAlle prop)







geefAlle :: Prop -> [Bedeling]
geefAlle = subs . (nubBy (eqList eqChar)) . geefVars



geefVars :: Prop -> Bedeling
geefVars (Var s) = [s]
geefVars (Bool _) = [""]
geefVars (p :-> q) = geefVars p ++ geefVars q
geefVars (En (p:ps)) = geefVars p ++ geefVars (En ps)
geefVars (Of (p:ps)) = (geefVars p) ++ (geefVars (Of ps))
geefVars (Niet p) = geefVars p
geefVars _ = []




subs :: [a] -> [[a]]
subs [] = [[]]
subs (x:xs) = map (x:) subsxs ++ subsxs
                  where subsxs = subs xs


















tautologie :: Prop -> Bool
tautologie p = t2 p (geefAlle p)




t2 :: Prop -> [Bedeling] -> Bool
t2 p [] = evalueer p []
t2 p (x:xs) | evalueer p x = t2 p xs
            | otherwise = False









contradictie :: Prop -> Bool
contradictie p = c2 p (geefAlle p)




c2 :: Prop -> [Bedeling] -> Bool
c2 p [] = not (evalueer p [])
c2 p (x:xs) | evalueer p x = False
            | otherwise = c2 p xs










equivalent :: Prop -> Prop -> Bool
equivalent p q = (e2 p (geefAlle p) q) && (e2 p (geefAlle q) q)





e2 :: Prop -> [Bedeling] -> Prop -> Bool
e2 p [] q     = eqBool (evalueer p []) (evalueer q [])
e2 p (x:xs) q | eqBool (evalueer p x) (evalueer q x) = e2 p xs q
              | otherwise = False









deMorgan :: Prop -> Prop
deMorgan (Var s)       = Var s
deMorgan (p :-> q)     = (deMorgan p) :-> (deMorgan q)
deMorgan (Of p)        | uitsluitendNiet (map deMorgan p) = Niet (En (map deMorgan (verwijderNiet (map deMorgan p))))
                       | otherwise         = Of (map deMorgan p)
deMorgan (En p)        | uitsluitendNiet (map deMorgan p) = Niet (Of (map deMorgan (verwijderNiet (map deMorgan p))))
                       | otherwise         = En (map deMorgan p)
deMorgan (Niet (En p)) = Of (map deMorgan (map Niet p))
deMorgan (Niet (Of p)) = En (map deMorgan (map Niet p))
deMorgan (Niet p)      = Niet (deMorgan p)
deMorgan (Bool b)      = Bool b




uitsluitendNiet :: [Prop] -> Bool
uitsluitendNiet []            = True
uitsluitendNiet ((Niet _):ps) = uitsluitendNiet ps
uitsluitendNiet (_)           = False




verwijderNiet :: [Prop] -> [Prop]
verwijderNiet []            = []
verwijderNiet ((Niet p):ps) = p : (verwijderNiet ps)
verwijderNiet (p:ps)        = p : (verwijderNiet ps)













toon :: Prop -> String
toon (Var s)     = s
toon (p :-> q)   = (toonHaakjes p) ++ " -> " ++ (toonHaakjes q)
toon (Of (p:ps)) | (length ps)==0 = toon p
                 | bevatIm p      = (toonHaakjes p) ++ " \\/ " ++ (toon(En ps))
                 | otherwise      = (toon p) ++ " \\/ " ++ (toon (Of ps))
toon (Of [])     = ""
toon (En (p:ps)) | (length ps)==0 && (bevatOf p) = toonHaakjes p
                 | (length ps)==0 = toon p
                 | bevatOf p || bevatIm p
                                  = (toonHaakjes p) ++ " /\\ " ++ (toon(En ps))
                 | otherwise      = (toon p) ++ " /\\ " ++ (toon (En ps))
toon (En [])     = ""
toon (Niet p)    = "!" ++ (toonHaakjes p) ++ ""
toon (Bool True) = "T"
toon (Bool False)= "F"




bevatOf :: Prop -> Bool
bevatOf (Of _) = True
bevatOf _      = False



bevatIm :: Prop -> Bool
bevatIm (_ :-> _) = True
bevatIm _         = False



toonHaakjes :: Prop -> String
toonHaakjes (Var s)     = s
toonHaakjes (p :-> q)   = "(" ++ (toon (p :-> q)) ++ ")"
toonHaakjes (Of (p:ps)) | (length ps)==0 = toon p
                        | otherwise      = "(" ++ (toon (Of (p:ps))) ++")"
toonHaakjes (Of [])     = ""
toonHaakjes (En (p:ps)) | (length ps)==0 = toon p
                        | otherwise      = "(" ++ (toon (En (p:ps))) ++")"
toonHaakjes (En [])     = ""
toonHaakjes (Niet p)    = "!" ++ (toonHaakjes p) ++ ""
toonHaakjes (Bool True) = "T"
toonHaakjes (Bool False)= "F"











ontleed :: String -> Prop

ontleed s | eqString s "T"          = Bool True
          | eqString s "F"          = Bool False
          | eqString s (zoekHaakjes s) = ontleed2 s
          | otherwise = Var (verwijderHaakjes 1 s)
                where verwijderHaakjes i t | i>(length t) = ""
                                           | (eqString (mid t i 1) "(") || (eqString (mid t i 1) ")")
                                                          = verwijderHaakjes (i+1) t
                                           | otherwise    = (mid t (i) 1) ++ (verwijderHaakjes (i+1) t)

ontleed2 :: String -> Prop
ontleed2 s | (zoek 1 s " /\\ ")>0
                       = En[(ontleed (zoekHaakjes (left s ((zoek 1 s " /\\ ")-1)))), (ontleed (zoekHaakjes (right s ((length s)-(zoek 1 s " /\\ ")-3))))]
           | (zoek 1 s " \\/ ")>0
                       = Of[(ontleed (zoekHaakjes (left s ((zoek 1 s " \\/ ")-1)))), (ontleed (zoekHaakjes (right s ((length s)-(zoek 1 s " \\/ ")-3))))]
           | (zoek 1 s " -> ")>0
                       = (ontleed (zoekHaakjes ((left s ((zoek 1 s " -> ")-1)))) :-> (ontleed (zoekHaakjes (right s ((length s)-(zoek 1 s " -> ")-3)))))
           | eqString (left s 1) "!"
                       = Niet (ontleed (zoekHaakjes (right s ((length s)-1))))
           | otherwise = "blabalbla "++s++"blablabal"






















zoekHaakjes :: String -> String
zoekHaakjes s = zhaak s 0 0 ""

zhaak :: String -> Int -> Int -> String -> String
zhaak [] _ _ s = s
zhaak (x:xs) ho hs s | ho == hs && ho>0 = s
                     | eqChar x '(' = if ho==0  then zhaak xs (ho+1) hs s
                                      else           zhaak xs (ho+1) hs (s++"(")
                     | eqChar x ')' = if ho==(hs+1) then zhaak xs ho (hs+1) s
                                      else               zhaak xs ho (hs+1) (s++")")
                     | ho > hs      = zhaak xs ho hs (s++[x])
                     | otherwise    = zhaak xs ho hs s




zoek :: Int -> String -> String -> Int
zoek i s t | i-1>(length s)-(length t)       = 0
           | eqString (mid s i (length t)) t = i
           | otherwise                       = zoek (i+1) s t




zoekEind :: Int -> String -> String -> Int
zoekEind i s t | i<=0                            = 0
               | eqString (mid s i (length t)) t = i-1
               | otherwise                       = zoek (i-1) s t




left :: String -> Int -> String
left [] _     = ""
left (s:ss) i | i>0       = s : (left ss (i-1))
              | otherwise = left ss (i-1)




right :: String -> Int -> String
right []     _ = ""
right (s:ss) i | (length ss)<=i-1  = s : (right ss (i))
               | otherwise         = right ss (i)




mid :: String -> Int -> Int -> String
mid s i l = right (left s (i+l-1)) l




cut :: String -> Int -> Int -> String
cut s i l = mid s i (l-i)









about :: IO()
about = do putStr ("/------------------------------------\\\n" ++
                   "|                                    |\n" ++
                   "|     [functioneel programmeren]     |\n" ++
                   "|                                    |\n" ++
                   "|     MATHIJS LAGERBERG   #0301523   |\n" ++
                   "|            and                     |\n" ++
                   "|      PAUL LAMMERTSMA    #0305235   |\n" ++
                   "|                                    |\n" ++
                   "|          . . present . .           |\n" ++
                   "|                                    |\n" ++
                   "|             -- the --              |\n" ++
                   "|     PROPOSITION THINGAMABOBBER     |\n" ++
                   "|                 ~                  |\n" ++
                   "|                                    |\n" ++
                   "\\------------------------------------/\n")



testProp :: Prop
testProp = (Var "p" :-> Var "q")

testString :: [Bedeling]
testString = [["p","q","r"],["p","q"],["p","q"],["p"],["q"],["p"],[]]

taut :: Prop
taut = (Of [(Of [Var "p", Niet (Var "q")]), (Of [Bool False, Bool True])])

contr :: Prop
contr = (En [Var "p", Niet (Var "p")])

-- should be: Var (...)
-- 278,26-53
