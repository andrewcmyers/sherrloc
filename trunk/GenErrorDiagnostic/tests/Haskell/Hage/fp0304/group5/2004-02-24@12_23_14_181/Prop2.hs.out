Compiling Prop2.hs
(278,26)-(278,54): Type error in guarded expression
 expression       : "blabalbla " ++ s ++ "blablabal"
   type           : [Char]
   does not match : Prop  

Compilation failed with 1 error
